File: Toggly.FeatureManagement.Java/toggly-core/src/main/java/io/toggly/core/model/VariantResult.java
package io.toggly.core.model;

import java.util.Objects;

/**
 * Assigned variant name, configuration value, and effective enabled state
 * for a feature, returned by {@code TogglyClient#getVariant}.
 *
 * <p>Catalog-local: assigned by the local MF-parity allocator
 * ({@code io.toggly.core.eval.VariantAllocator}) directly from
 * {@code FeatureDefinition#getVariants()} / {@code getAllocation()} — never
 * fetched separately from the definitions payload.</p>
 *
 * <p>{@link #isEnabled()} reflects the feature's effective enabled state
 * <em>after</em> applying the assigned variant's
 * {@link VariantStatusOverride}, matching
 * {@code Microsoft.FeatureManagement}'s {@code GetVariantAsync} semantics.
 * It may differ from {@code TogglyClient#isEnabled}, which stays purely
 * filter-based and is never affected by {@code StatusOverride}.</p>
 */
public final class VariantResult {

    private final String name;
    private final Object configurationValue;
    private final boolean enabled;

    /**
     * Creates a variant result.
     *
     * @param name assigned variant name
     * @param configurationValue configuration payload for the variant
     * @param enabled effective enabled state after {@code StatusOverride}
     */
    public VariantResult(String name, Object configurationValue, boolean enabled) {
        this.name = name;
        this.configurationValue = configurationValue;
        this.enabled = enabled;
    }

    /**
     * Variant name assigned by the local allocator.
     */
    public String getName() {
        return name;
    }

    /**
     * Configuration payload for the variant.
     */
    public Object getConfigurationValue() {
        return configurationValue;
    }

    /**
     * Effective enabled state after applying the variant's
     * {@link VariantStatusOverride}, if any.
     */
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        VariantResult that = (VariantResult) o;
        return enabled == that.enabled
                && Objects.equals(name, that.name)
                && Objects.equals(configurationValue, that.configurationValue);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, configurationValue, enabled);
    }

    @Override
    public String toString() {
        return "VariantResult{" +
                "name='" + name + '\'' +
                ", configurationValue=" + configurationValue +
                ", enabled=" + enabled +
                '}';
    }
}

File: toggly-ruby/toggly/lib/toggly/variant_result.rb
# frozen_string_literal: true

module Toggly
  # Assigned variant for a feature, returned by `Client#get_variant`.
  #
  # Assignment is computed locally from the feature's catalog
  # `variants` / `allocation` (see {VariantAllocator}), matching
  # `Microsoft.FeatureManagement`'s `IVariantFeatureManager` bit-for-bit.
  # There is no network round-trip: this is not the `enable_variants`
  # dual-rail from earlier releases (removed; see CHANGELOG).
  class VariantResult
    # @return [String] Variant name
    attr_reader :name

    # @return [Object, nil] Untyped configuration payload for this variant
    attr_reader :configuration_value

    # @return [Boolean] Effective enabled flag for this feature/context after
    #   the variant's `StatusOverride` is applied. This can differ from
    #   `Client#enabled?`, which stays filter-based only; use this field when
    #   you need MF-identical `GetVariantAsync`-style effective-enabled
    #   semantics (e.g. a variant with `StatusOverride: "Enabled"` on an
    #   otherwise-disabled feature).
    attr_reader :enabled

    # @return [String] Assignment reason: "User" | "Group" | "Percentile" |
    #   "DefaultWhenEnabled" | "DefaultWhenDisabled"
    attr_reader :reason

    def initialize(name:, configuration_value: nil, enabled: true, reason: nil)
      @name = name
      @configuration_value = configuration_value
      @enabled = enabled
      @reason = reason
    end

    # @return [Hash]
    def to_h
      { name: @name, configuration_value: @configuration_value, enabled: @enabled, reason: @reason }
    end

    def ==(other)
      other.is_a?(VariantResult) &&
        @name == other.name &&
        @configuration_value == other.configuration_value &&
        @enabled == other.enabled &&
        @reason == other.reason
    end
    alias eql? ==
  end
end
